README
README for pathrate-2.4.1

Contents
--------
Overview of pathrate
How pathrate works
Building pathrate
Running pathrate
Examples
Changes from previous pathrate release
Contact and other info


Overview of pathrate
--------------------
pathrate measures the capacity of a network path from host S (sender) to host R
(receiver). The capacity of a path, aka bottleneck bandwidth, is the maximum
IP-layer throughput that a flow can get in the path from S to R. The capacity
does not depend on the load of the path. 

* The capacity of a path is not the same metric as the available bandwidth
of a path. The later takes into account the cross traffic load in the path, and
it is less than the capacity. Also, the capacity is different than the Bulk
Transfer Capacity (BTC) metric, currently standardized by the IETF. The BTC is
the maximum throughput that a TCP connection can get in the path, when it is
only limited by the network.

* pathrate requires that you have access at both end-points of the 
network path (i.e., you have to be able to run pathrate at both S and R).  This
makes the tool harder to use, but it is the only way to avoid distorting the
measurements with the reverse path from R to S. 

* You can run pathrate from user-space, and you don't need any superuser 
privileges.

* Pathrate takes normally about 15 minutes to run. Even though we could
simplify the measurement process so that it takes much shorter, we chose to
focus on accuracy rather than on execution speed.  Notice that the capacity of
a path is a static metric that does not change unless if there are routing or
infrastructure changes in the path.  Consequently, the long execution time of
pathrate should not be a concern. 

* It is important to run pathrate from relatively idle hosts. Before 
running pathrate, make sure that there are no other CPU or I/O intensive
processes running. If there are, it is likely that they will interact with
pathrate's user-level packet timestamping, and the results that you'll get may
be quite inaccurate.

* Certain links use load balancing (e.g., Cisco's standard CEF load-sharing). 
In those cases, even though a certain "fat link" may have a capacity X, an IP
flow will only be able to see a maximum bandwidth of X/n, where n is the number
of "sub-links" (e.g., ATM PVCs) that make up the "fat link". 

* In paths that are limited by Gigabit Ethernet interfaces,
the Pathrate final capacity estimate is often less than 1000Mbps. The major
issue there is whether the two end-hosts can truly saturate a Gigabit Ethernet
path, even when they transfer MTU UDP packets.  We have observed that in order
to saturate a GigE interface, it is important that end-hosts have a 64-bit PCI
bus (together with a GHz processor and a decent GigE NIC of course).  In other
words, Pathrate cannot measure a nominal network capacity if the end-hosts are
not really able to use that capacity.

* Some links perform traffic shaping, providing a certain
peak rate P, while if the burst size is larger than a certain amount of bytes,
the maximum rate is reduced to a lower ("sustainable") rate S. In such paths,
Pathrate should measure P, not S.  Thanks to Tom Dunigan from ORNL that run
Pathrate in such paths and gave us feedback.

* In previous versions, Pathrate would sometimes report that it
cannot produce a final capacity estimate (especially, in      high-bandwidth
paths). In the new version, Pathrate will report instead a lower-bound on the
capacity of a path. 

* Internet paths are often asymmetric. The capacity of the path from S to R 
is not necessarily the same with the capacity of the path from R to S.

* For heavily loaded paths, pathrate can take a while until it reports a 
final estimate. "A while" means about half an hour. The good news is that the
capacity of a path does not change very often, unless if there is route
flapping. 

* pathrate uses UDP packets for probing the path's bandwidth, and it also 
establishes a TCP connection between the two hosts for control purposes.  The
UDP port number is 48698 (at the receiver) and the TCP port number is 48699 (at
the sender).

* pathrate does some primitive form of congestion avoidance (it will abort
after many packet losses).

* pathrate assumes that the IP and UDP headers (28 bytes totally) are
fully transmitted together with the packet payload. For links that do header
compression (RFC 1144) this will cause a slight capacity overestimation.  
 
* pathrate v2.x is significantly different than the first version, released 
in November 2000. Also, pathrate v2.x differs in certain points with the
description given in our Infocom 2001 paper.



How pathrate works
------------------
Pathrate consists of the following "run phases":

1) Initially, the tool discovers the maximum train-length that the path can
carry. The idea is that we do not want to overload the path with very long
packet trains that would cause buffer overflows. The maximum train length that
we try is 50 packets. We stop increasing the train length after three lossy
packet trains at a given train length.

2) Then, pathrate sends a number of packet trains of increasing length (called
"preliminary measurements" phase). The goal here is to detect if the narrow
link has parallel sub-channels, or if it performs traffic shaping. You can
ignore this phase until you become an "advanced user". This phase also checks
whether the path is "easy to measure" (very lightly loaded). In that case,
pathrate reports its final estimate and exits. An important part of this phase
is that pathrate computes the "bandwidth resolution" (think of it as a
histogram bin width). The final capacity estimate will be a range of this
width.

3) In this phase, called Phase I, pathrate generates a large number (1000) of
packet-pairs. The goal here is to discover all local modes in the packet-pair
bandwidth distribution. One of the Phase I modes is expected to be the capacity
of the path. The packets that pathrate sends in Phase I are of variable size,
in order to make the non-capacity local modes weaker and wider. 

4) Finally, in Phase II, pathrate generates a number (500) of packet trains of
large length. The goal here is to discover which of the Phase I local modes is
the capacity of the path. To do so, Phase II estimates the Asymptotic
Dispersion Rate (ADR) metric R, measured from the dispersion of long packet
trains. We know (see our Infocom 2001 paper) that the capacity is larger than
R. If there are several Phase I modes that are larger than R, the capacity
estimate is the mode that maximizes a certain "figure of merit" M. M depends on
the "narrowness" and the "strength" of each candidate mode in the underlying
bandwidth distribution of Phase I. The capacity mode should be narrow and
strong, i.e., to have a large value of M. 

The very final outcome of pathrate is the capacity estimate for the path. 




Building pathrate
-----------------
pathrate uses the standard configure/make approach.
$> ./configure
$> make

After you have extracted the pathrate code in a directory,
run configure in that directory. Then run make.

You should then have two executables in the pathrate directory.
1. pathrate_snd (to be run at the sender)
2. pathrate_rcv (to be run at the receiver)



Running pathrate
----------------

sender 
   For capacity estimation, first run pathrate_snd at the sender S:
   $> pathrate_snd
   By default, pathrate_snd will exit after 1 measurement.
   Use switch -i to run sender in iterative mode.
   Other options are,
        -q        : quite mode
        -v        : verbose mode
        -o <file> : print log in user specified file
        -H|-h     : print help and exit                                                                                       

receiver 
   Then, run pathrate_rcv at the receiver R:
   $> pathrate_rcv [-q|-v] [-Q] [-o|-O <filename>] [-N <filename>] [-H] -h <sender>
        -s        : hostname/ipaddress of sender
        -q        : quite mode
        -v        : verbose mode
        -Q        : Quick termination mode
        -o <file> : print log in user specified file
        -O <file> : append log in user specified file [default is pathrate.output]
        -N <file> : print output in ULM format to <file>
        -H|-h     : print help and exit
      
   [-o,-O] By default, receiver always appends detailed output to file 
   "pathrate.output". output file can be changed by specifying a filename with -o 
    switch. Remember, if the specified file already exits, it will be over-written.
    To append the output to a file use -O switch.
   
   [-q,-v] By default, receiver sends minimal output (Progress report, Local 
   modes and final output) to terminal (stdout). To completely disable any 
   message on stdout, please use -q (quite mode).
   To enable detailed output on screen, use -v (Verbose mode). Note that, log file 
   (user specified or "pathrate.output") contains exactly the same output as -v.
   
   [-N] Specify a file name to store output in netlogger compatible format.

   [-Q] If used, terminates at the end of phase 0 and give rough estimate of
   capacity by getting the average of the estimates so far. It also reports
   the coefficient of variation (standard deviation/ average). If this variation
   is large, complete run of pathrate is suggested for more accurate capacity
   estimate.



Examples
--------
Say, you want to measure the capacity from host A to host B, then run 
pathrate_snd at A and pathrate_rcv at B.

At host A,
$> ./pathrate_snd -i -q &

At host B,
$> ./pathrate_rcv -s A 

Now, if you wanted to change the default log file at receiver B,
$> ./pathrate_rcv -s A -o testrun.log

Next if you also wanted detailed output on screen and netlogger
output, 
$> ./pathrate_rcv -s A -o testrun.log -N netlogger.log -v


Changes from previous pathrate release
--------------------------------------
1) A command line options are added for users to choose different levels 
   of verbose output on the screen. It also has options for redirecting
   output to different files and appending to file instead of 
   overwriting.

2) Current version has been tested to work at high capacity paths.
   (OC-3, OC-12, Gigabit). It can detect interrupt coalescing at
   Gigabit NICs and adopt different methodology for capacity 
   estimation.

3) It can report final estimation in Universal Logger Message (ULM)
   format.

Authors
-------
1. Constantinos Dovrolis
   dovrolis@cis.udel.edu
2. Ravi S Prasad 
   ravi@cc.gatech.edu


Important Disclaimer
--------------------
pathrate is based on statistical techniques, and so there is no guarantee that
it will always estimate the correct capacity value. If you get wrong results, 
please let us know, including also the corresponding <pathrate.output> file.

Please report any problems or suggestions to dovrolis@cc.gatech.edu and 
ravi@cc.gatech.edu. Also, please let us know if you want to receive email 
announcements about future pathrate releases. 


Acknowledgments
---------------
This work was supported by the SciDAC program of the US Department 
of Energy (award # DE-FC02-01ER25467).


References 
----------
  "What do packet dispersion techniques measure?",
      C. Dovrolis, P. Ramanathan, D.Moore, 
      Proceedings of IEEE Infocom 2001 
      http://www.cc.gatech.edu/~dovrolis/Papers/infocom01.ps
and
  "Packet dispersion techniques and capacity estimation",
      C. Dovrolis, P. Ramanathan, D.Moore,
            submitted for publication, 
            http://www.cc.gatech.edu/~dovrolis/Papers/ton_dispersion.ps


CHANGES
Changes in 2.4.1 (Jan 2005)
---------------------------
* Modified the sender receiver interaction during retransmission 
  of a lost train. Now receiver sends SEND signal retransmission
  also instead of sender retransmitting in response to NEG_ACK

* Added some debug messages that can be activated by adding -DDEBUG in 
  the makefile

Changes in 2.4.0 (Apr 2004)
---------------------------
* Sender - receiver interaction
  Receiver requests each train and sends ACK or negative ACK.
  Now the receiver will not work with older version of the sender.
* Interrupt coalescence changes
  Detects IC and uses alternate algorithm (Published in PAM '04).
* Minor bug fixes
  - Corrected option (-o <file>) implementation in both sender and receiver

Changes in 2.3.3 (Jan 2004)
---------------------------
* Report ADR as capacity mode only when there is no mode larger than ADR
  (Sometimes ADR mode is narrow even when there exist correct caapcity mode.)

* Bug fixes.
  - Changed exit condition in recv_train function from 
      while ((pack_id<train_len )
    to while ((exp_pack_id<=train_len ); was affecting results in lossy networks.
  - Frees most of the memory allocation now.
  - Slightly modified interrupt coalesence detection.
  - Few print anamolies in gigabit test function.

Changes in 2.3.2 (Jun 2003)
---------------------------
Weiling's changes 2:
* kurtosis instead of ratio of mode_cnt and bell_cnt is used as the narrowness of the bell
  to calculate figure of merit.
* BIN_NOISE_TOLER is modified to be proportional to previous bin_cnt
  instead of a constant.

Changes in 2.3.1 (May 2003)
---------------------------
Weiling's changes 1:
* bin_wd is increased to be twice of the value used earlier
* bell_cnt and index are updated without the previously required condition that
  bin has significant number of measurements.

Changes in 2.3.0 (April 2003):
-----------------------------
* Tested on high capacity paths. (OC-3, OC-12 and Gigabit paths)

* Detects interrupt coalescing. If detected, uses long trains to 
  estimate capacity. (tested on Intel Pro 1000 Gigabit controlor
  82540)

* added command line options to
  - suppress output (-q)
  - get more verbose output (-v)
  - get quick estimate (-Q)
  - write/append output to a file (-o/O)
  - append final outoput in ULM format to a file (-N)

* Receiver uses different thread for control messages while getting probing
  packets in main thread. Don't use select anymore.

* recv_train function return array of packet arrival times.
  
* Sender performs sleep time check before waiting for connection
  from the reciever. Once per sender run, instead of once per connection.

* Abort phase 1 when train length > MAX(max_train_len/4,2)
        rather than when > MIN(max_train_len, 12)

* Fixed the number of pack sizes (to 40) rather than the increament in 
        the packet size (was 25B)

* Fixed small printing bugs

Changes in 2.2.1 (April 2002): 
------------------------------
* limit max train length in PHASE I.
  -  Aborts Phase I 
     when train length > 12. // It's packet-train now and give ADR measurement.

* Dealing with ignored msrments in Phase I changed to:
	Recv pairs/trains in blocks of 27
	At the end of each "block", 
	   if #of ignored msrments > IGNORED_LIM*27
		train_len+=1;
		pack_sz+=200

  ---- fixed small printing bugs.
  ---- Previous version was printing wrong packet sizes.

To-do list :
*) Find a "correct" threshold factor for ignoring msrments.	
*) Find a better binwd


Changes in 2.2.0 (April 2002): 
------------------------------
* select timeout made compatible with Linux 		  	| in _snd.c
* Sender sends abort signal to rcvr to end measurements 
	if it sees too many losses     				| in _snd.c
* Use select in recv_train routine on both the 
	data and control sockets.				| in _rcv.c
* Check TCP socket in recv_train if no data in UDP socket
	Presently and data on TCP socket while expecting a 
	probing packet implies an abort message.		| in _rcv.c
* If(No Phase I mode > ADR && Phase II != unimodal) 
	Then capacity = max_merit(Phase II modes)		| in _rcv.c
* Pathoverflow => go back to prev-train-length && don't change
			train spacing				| in _rcv.c
* Dealing with ignored msrments in Phase I:
	Recv pairs/trains in blocks of 27
	At the end of each "block", 
	   if #of ignored msrments > IGNORED_LIM*27
		train_len+=2;
		pack_sz+=200
	   If train_len > max_train_len	
		train_len=maxtrain_len				| in _rcv.c
* No such measures in Phase II.
* New bin width :
   - 	based on max-min (after removing 10% outliers) 
	instead of ave. 					
   -	Simpler fraction of bw_range				| in _rcv.c
* Preliminary measurements 
   - # of outliers is fraction of measured 			| in _rcv.c
* messages :
   - Show "time" remaining to complete.


To-do list :
*) Find a "correct" threshold factor for ignoring msrments.	
*) Find a better binwd



pathrate.h
/*
 This file is part of pathrate.

 pathrate is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License, or
 (at your option) any later version.

 pathrate is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with pathrate; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

/*-------------------------------------------------
   pathrate : an end-to-end capcity estimation tool
   Author   : Constantinos Dovrolis (dovrolis@cc.gatech.edu )
              Ravi S Prasad            ( ravi@cc.gatech.edu )
              
   Release  : Ver 2.4.1
   Support  : This work was supported by the SciDAC
              program of the US department 
--------------------------------------------------*/


#ifdef LOCAL
#define EXTERN
#else
#define EXTERN extern
#endif

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <arpa/inet.h>
#include <errno.h>
#include <ctype.h>
#include <netdb.h>
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <strings.h>
#include <pthread.h>
#include <string.h>
#include <unistd.h>
#include <sys/time.h>
#include <time.h>
#include <math.h>
#include <float.h>
#include <fcntl.h>

#define MAX_V(a,b)	((a)>(b)? (a): (b))
#define MIN_V(a,b)	((a)<(b)? (a): (b))

typedef struct {
  double mode_value_lo;	/* Lower bandwidth value (Mbps) of local mode */
  double mode_value_hi;	/* Upper bandwidth value (Mbps) of local mode */
	long   mode_cnt;	/* number of measurements in local mode */ 
	long   bell_cnt;	/* number of measurements in "bell" of local mode */ 
  double bell_lo;		/* low threshold (Mbps) of bell */
	double bell_hi;		/* high threshold (Mbps) of bell */  
	double bell_kurtosis;		  
} mode_struct;

#define  UNIMPORTANT_MODE 	-1
#define  LOCAL_MODE 	1
#define  LAST_MODE 		-2

/* Code numbers sent from receiver to sender using the TCP control stream */
#define  TRAIN_LEN        	0x00000001
#define  NO_TRAINS      		0x00000002
#define  GAME_OVER         	0x00000003
#define  ACK_TRAIN        	0x00000004
#define  PCK_LEN          	0x00000005
#define  MAX_PCK_LEN       	0x00000006
#define  SEND			          0x00000007
#define  CONTINUE	        	0x00000008
#define  TRAIN_SPACING  		0x00000009
#define  SENT_TRAIN     		0x00000010
#define  NEG_ACK_TRAIN    	0x00000011

/* Port numbers (UDP for receiver, TCP for sender) */
#define  UDPRCV_PORT 		48698 
#define  TCPSND_PORT 		48699 

#define  UDP_BUFFER_SZ  	212536 

#define  MAX_CONSEC_LOSSES	30
#define  MAX_PACK_SZ    	16384

#define  MIN_TRAIN_SPACING	500000 	/* microseconds */	
#define  TRAIN_SPACING_SEC	1	/* seconds */

EXTERN long verbose ;
EXTERN long Verbose ;
EXTERN int sock_tcp, sock_udp;


/*
 *  * Weiling: BIN_CNT_TOLER_kernel_percent */
#define  BIN_CNT_TOLER_kernel_percent   0.1



pathrate_snd.h
/*
 This file is part of pathrate.

 pathrate is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License, or
 (at your option) any later version.

 pathrate is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with pathrate; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

/*-------------------------------------------------
   pathrate : an end-to-end capcity estimation tool
   Author   : Constantinos Dovrolis (dovrolis@cc.gatech.edu )
              Ravi S Prasad            ( ravi@cc.gatech.edu )
              
   Release  : Ver 2.4.1
   Support  : This work was supported by the SciDAC
              program of the US department 
--------------------------------------------------*/


#ifdef LOCAL
#define EXTERN
#else
#define EXTERN extern
#endif

EXTERN int iterative;
EXTERN long recv_ctr_msg(int ctr_strm, char *ctr_buff) ;
EXTERN void send_ctr_msg(int ctr_strm, long ctr_code) ;
EXTERN double time_to_us_delta(struct timeval tv1, struct timeval tv2);
EXTERN void help(void);


pathrate_snd_func.c
/*
 This file is part of pathrate.

 pathrate is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License, or
 (at your option) any later version.

 pathrate is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with pathrate; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

/*-------------------------------------------------
   pathrate : an end-to-end capcity estimation tool
   Author   : Constantinos Dovrolis (dovrolis@cc.gatech.edu )
              Ravi S Prasad            ( ravi@cc.gatech.edu )
              
   Release  : Ver 2.4.1
   Support  : This work was supported by the SciDAC
              program of the US department 
--------------------------------------------------*/


#include "pathrate.h"
#include "pathrate_snd.h"

/* 
	Receive a message from the control stream 
*/
long recv_ctr_msg(int ctr_strm, char *ctr_buff) 
{ 
	long ctr_code;
  if (read(ctr_strm, ctr_buff, sizeof(long)) != sizeof(long)) 
	  return(-1);
	memcpy(&ctr_code, ctr_buff, sizeof(long)); 
	return(ntohl(ctr_code));
}

/*
    Send an empty message to the control stream
*/
void send_ctr_msg(int ctr_strm, long ctr_code) 
{
  char ctr_buff[8];
  long ctr_code_n = htonl(ctr_code);
  memcpy(ctr_buff, &ctr_code_n, sizeof(long));
  if (write(ctr_strm, ctr_buff, sizeof(long)) != sizeof(long)) {
    fprintf(stderr, "send control message failed:\n");
    exit(-1);
  }
}

/*
    Compute the time difference in microseconds between two timeval measurements
*/
double time_to_us_delta(struct timeval tv1, struct timeval tv2)
{
	double time_us;
	time_us= (double) ((tv2.tv_sec-tv1.tv_sec)*1000000 + (tv2.tv_usec-tv1.tv_usec));
	return time_us;
}

/*
 *  Help
 *  */
void help(){
  fprintf (stderr,"pathrate_snd options\n");
  fprintf (stderr,"-i        : iterative mode\n");
  fprintf (stderr,"-q        : quite mode\n");
  fprintf (stderr,"-v        : verbose mode\n");
  fprintf (stderr,"-o <file> : print log in file\n");
  fprintf (stderr,"-H|-h     : print this help and exit\n");
}


pathrate_rcv.h
/*
 This file is part of pathrate.

 pathrate is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License, or
 (at your option) any later version.

 pathrate is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with pathrate; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

/*-------------------------------------------------
   pathrate : an end-to-end capcity estimation tool
   Author   : Constantinos Dovrolis (dovrolis@cc.gatech.edu )
              Ravi S Prasad            ( ravi@cc.gatech.edu )
              
   Release  : Ver 2.4.1
   Support  : This work was supported by the SciDAC
              program of the US department 
--------------------------------------------------*/


#ifdef LOCAL
#define EXTERN
#else
#define EXTERN extern
#endif

#define  MAX_TRAIN_LEN      50
#define  MAX_IGNORED_MSRMS	0.3
#define  IGNORE_LIM_FACTOR	4

#define  MIN_PACK_SZ_P1		572 
#define  SIZE_INCR_STEP	 	25	

#define  NO_TRAINS_P1		1000
#define  NO_TRAINS_P2		500

#define  COEF_VAR_THR 		0.05
#define  ADR_REDCT_THR 		0.90

#define  BIN_NOISE	   	15
#define  BIN_CNT_TOLER		5	
#define  MAX_NO_MODES	 	1000

EXTERN int sock_tcp, sock_udp;
EXTERN char message[1024];
EXTERN FILE *pathrate_fp;	
EXTERN FILE *netlog_fp;	
EXTERN int netlog;
EXTERN char hostname[256];
EXTERN int recv_train_done;
EXTERN int retransmit;

EXTERN void prntmsg(FILE *fp);
EXTERN void help(void);
EXTERN void send_ctr_msg(long ctr_code) ;
EXTERN long recv_ctr_msg(int ctr_strm, char *ctr_buff);
EXTERN void print_bw(FILE *fp, double bw_v) ;
EXTERN void termint(int exit_code);
EXTERN void happy_end(double bw_lo, double bw_hi);
EXTERN double time_to_us_delta(struct timeval tv1, struct timeval tv2) ;
EXTERN void time_copy(struct timeval time_val_old, struct timeval *time_val_new) ;
EXTERN void order(double unord_arr[], double ord_arr[], long no_elems);
EXTERN double get_avg(double data[], long no_values);
EXTERN double get_std(double data[], long no_values);
/*Weiling: *bell_kurtosis*/
EXTERN double get_mode(double ord_data[], short vld_data[], double bin_wd, 
       int no_values, mode_struct *curr_mode);
EXTERN void sig_alrm(int signo);
EXTERN void *ctr_listen(void *arg);
EXTERN int recv_train(int train_len, int *round, struct timeval *time[]) ;
EXTERN int gig_path(int pack_sz, int round, int k_to_u_latency);
EXTERN void print_time(FILE *fp, int time);
EXTERN int intr_coalescence(struct timeval time[],int len, double latency);
